articles/actions.md

```diff
@@ -0,0 +1,338 @@
+After a goal has been accepted the client should call this service to receive the result.
+The result will indicate the final status of the goal and any user defined data.
+
+Once the server sends the result to the client it should free up any resources used by the action.
+If the client never asks for the result then the server should discard the result after a timeout period.
+
+### Goal Status Topic
+
+* **Direction**: Server publishes
+* **Content**: List of in-progress goals with: Goal ID, time accepted, and an enum indicating the status
+
+This topic is published by the server to broadcast the status of goals it has accepted.
+The purpose of the topic is for introspection; it is not used by the action client.
+Messages are published when transitions from one status to another occur.
+
+The default QoS settings for a DDS middleware should be TRANSIENT LOCAL with a history size of 1.
+This allows new subscribers to always get the latest state.
+
+The possible statuses are:
+
+* *Accepted*
+  * The goal has been accepted by the action server and may now be executing
+* *Cancelling*
+  * The action server will try to cancel the indicated goal
+* *Cancelled*
+  * The action server successfully canceled the goal
+* *Succeeded*
+  * The action server successfully reached the goal
+* *Aborted*
+  * The action server failed reached the goal
+
+### Feedback Topic
+
+* **Direction**: Server publishes
+* **Content**: Goal id, user defined feedback message
+
+This topic is published by the server to send application specific progress about the goal.
+It is up to the author of the action server to decide how often to publish the feedback.
+
+### Client/Server Interaction Examples
+
+Here are a couple of sequence diagrams depicting typical interactions between an action client and action server.
+
+#### Example 1
+
+In this example, the action client request a goal and gets a response from the server accepting the goal (synchronous).
+Upon accepting the goal, the action server starts a user defined execution method for completing the goal.
+Following the goal request, the client makes an asynchronous request for the result.
+The user defined method publishes feedback to the action client as it executes the goal.
+Ultimately, the user defined method populates a result message that is used as part of the result response.
+
+![Goal Lifecycle Example 0](../img/actions/interaction_example_0.png)
+
+#### Example 2
+
+This example is almost identical to the first, but this time the action  client requests for the goal to be canceled mid-execution.
+Note that the user defined method is allowed to perform any shutdown operations after the cancel request before returning with the cancellation result.
+
+![Goal Lifecycle Example 1](../img/actions/interaction_example_1.png)
+
+#### Example 3
+
+Here is a more complex example involving multiple goals.
+
+TODO
+
+## Bridging between ROS 1 and ROS 2
+
+TODO
```

Comment:
?

Reply:
Some content was added, but then deleted after receiving feedback.

Actions across the bridge aren't targeted for Crystal ( see ros2/ros2#583 ) so the section was deleted rather than updated. Here is the old content in case someone is curious or interested in writing this section.

<details><summary>Deleted content for `Bridging between ROS 1 and ROS 2`</summary>

```markdown
## Bridging between ROS 1 and ROS 2

### Detecting Action Servers and Clients

There will be an API for the bridge to get a list of all ROS 2 action servers and clients.

ROS 1 action servers and clients can be discovered by looking for topic ending in `/status` with message type `actionlib_msgs/GoalStatusArray`.
If a publisher exists then an action server exists.
If a subscriber exists then a client exists.

### Bridging ROS 1 Action Client and ROS 2 Action Server

If either a ROS 2 action server or a ROS 1 action client exists then the ROS 1 bridge will create the following:

1. a ROS 1 action server
2. a ROS 2 action client

#### Goal submission

When the ROS 1 action client submits a goal the bridge will check if a ROS 2 action server exists.
If no server exists then the goal is rejected, otherwise the bridge calls the ROS 2 goal submission service.
If the ROS 2 server accepts the goal then the bridge will call `setAccepted` on the ROS 1 bridge, otherwise it will call `setRejected`.

When a goal is accepted the bridge stores a map between the ROS 1 and ROS 2 goal IDs.
If a client submits a goal with the same ID then it will be rejected until the first goal is finished.

#### Goal Cancellation

When a ROS 1 client tries to cancel a goal the bridge must try to cancel it on the ROS 2 server.
First the bridge must look up the mapping of ROS 1 to ROS 2 goal IDs.
If a mapping exists then the bridge will call the cancellation service.

If the mapping does not exist because the server has not yet accepted or rejected the goal then the bridge will wait until the goal is accepted.
If it is rejected the bridge will call setRejected and ignore the cancellation request.
If it is accepted then it will call the cancellation service.

#### Feedback

The bridge will only publish feedback messages for goals that were sent through it.
Since the goal ID is generated by the server, the bridge can only know the mapping of ROS 1 to ROS 2 goal IDs if the bridge sent the goal.

#### Result

Once a goal has been accepted the bridge will wait for the result.
When it gets the result the bridge will call

If the bridge notices the ROS 2 action server disappears from the node graph then bridge will wait for the result indefinitely.
When the bridge reappears the bridge will try to get the result from the goal again.
If the ROS 2 server does not know of the goal, then the bridge will notify the ROS 1 client that the goal was canceled.

### Bridging ROS 1 Action Server and ROS 2 Action Client

If either a ROS 1 action server or a ROS 2 action client exist then the bridge will create:

1. a ROS 1 action client
2. a ROS 2 action server

#### Goal submission

When the ROS 2 action client submits a goal the bridge will check if a ROS 1 action server exists.
If it does not exist the bridge will reject the goal, otherwise it will submit the goal to the ROS 1 server.
The same goal ID is to be used for both ROS 1 and ROS 2.

Goals sent to the bridge from ROS 2 are always immediately accepted by the bridge.
The bridge will then submit the goal to the ROS 1 server.
The reason for instantly accepting the goal is that if the ROS 1 action server never responds to a goal request then the bridge cannot return a result.
Goals that are rejected by ROS 1 action servers will be reported as canceled to ROS 2 action clients.

#### Goal Cancellation

When a ROS 2 client tries to cancel a goal the bridge will immediately accept the cancellation request.
It will then try to cancel the request on the ROS 1 action server.
If the cancellation request is rejected by the ROS 1 server then the ROS 2 bridge will stay in the *CANCELING* state until the result of the goal is known.

#### Feedback

Since the goal ID is the same for both ROS 1 and ROS 2, the bridge will always publish feedback from ROS 1 servers to the ROS 2 feedback topic.

#### Result

When the ROS 1 action server publishes a result it will be set as the result on the ROS 1 bridge.
If the ROS 2 action client never calls the service to get the result then it is subject to the same timeout as if it were a normal ROS 2 action server.
```
</details>

<br>
One issue is it has the bridge always accepting goals from ROS 2 action clients before contacting the ROS 1 action server, but that hides what's happening on the ROS 1 action server from ROS 2 introspection tools. Another issue is it "detects" a ROS 1 action server or client using only a single topic, but servers or clients aren't actually functional until all topics are in use. Yet another issue is it always creates an action server if there is an action client on one side, which makes it seem like there is an action server available when there isn't.

Reply:
why not include this so the effort is not lost, but make clear its not targeted for December

Reply:
That's a good point. I want to avoid someone spending their time reviewing it since it has many known issues already, so I opened ros2/design#195 instead.